be/src/exec/table_function_node.cpp

```diff
@@ -0,0 +1,341 @@
+// eg: there are 3 functions in `_fns`
+//      eos:    false, true, true
+//      return: 1
+//
+//      eos:    false, false, true
+//      return: 2
+//
+//      eos:    false, false, false
+//      return: -1
+//
+//      eos:    true, true, true
+//      return: 0
+//
+// return:
+//  0: all fns are eos
+// -1: all fns are not eos
+// >0: some of fns are eos
+int TableFunctionNode::_find_last_fn_eos_idx() {
+    for (int i = _fn_num - 1; i >=0; --i) {
+        if (!_fns[i]->eos()) {
+            if (i == _fn_num - 1) {
+                return -1;
+            } else {
+                return i + 1;
+            }
+        }
+    }
+    // all eos
+    return 0;
+}
+
+// Roll to reset the table function.
+// Eg:
+//  There are 3 functions f1, f2 and f3 in `_fns`.
+//  If `last_eos_idx` is 1, which means f2 and f3 are eos.
+//  So we need to forward f1, and reset f2 and f3.
+bool TableFunctionNode::_roll_table_functions(int last_eos_idx) {
+    bool fn_eos = false;
+    int i = last_eos_idx - 1;
+    for (; i >= 0; --i) {
+        _fns[i]->forward(&fn_eos);
+        if (!fn_eos) {
+            break;
+        }
+    }
+    if (i == -1) {
+        // after forward, all functions are eos.
+        // we should process next child row to get more table function results.
+        return false;
+    }
+
+    for (int j = i + 1; j < _fn_num; ++j) {
+        _fns[j]->reset();
+    }
+
+    return true;
+}
+
+Status TableFunctionNode::get_next(RuntimeState* state, RowBatch* row_batch, bool* eos) {
+    RETURN_IF_ERROR(exec_debug_action(TExecNodePhase::GETNEXT));
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+
+    const RowDescriptor& parent_rowdesc = row_batch->row_desc();
+    const RowDescriptor& child_rowdesc = _children[0]->row_desc();
+    if (_parent_tuple_desc_size == -1) {
+        _parent_tuple_desc_size = parent_rowdesc.tuple_descriptors().size();
+        _child_tuple_desc_size = child_rowdesc.tuple_descriptors().size();
+        for (int i = 0; i < _child_tuple_desc_size; ++i) {
+            _child_slot_sizes.push_back(child_rowdesc.tuple_descriptors()[i]->slots().size());
+        }    
+    }
+
+    uint8_t* tuple_buffer = nullptr;
+    Tuple* tuple_ptr = nullptr;
+    Tuple* pre_tuple_ptr = nullptr;
+    int row_idx = 0;
+
+    while (true) {
+        RETURN_IF_CANCELLED(state);
+        RETURN_IF_ERROR(state->check_query_state("TableFunctionNode, while getting next batch."));
+
+        if (_cur_child_batch == nullptr) {
+            _cur_child_batch.reset(new RowBatch(child_rowdesc, state->batch_size(), mem_tracker().get()));
+        }
+        if (_child_batch_exhausted) {
+            // current child batch is exhausted, get next batch from child
+            RETURN_IF_ERROR(_children[0]->get_next(state, _cur_child_batch.get(), eos));
+            if (*eos) {
+                break;
+            }
+            _cur_child_offset = 0;
+            RETURN_IF_ERROR(_process_next_child_row());
```

Comment:
Can these lines be omitted?

Reply:
It need to be called before we call `_find_last_fn_eos_idx()` at first time